File: commands/commands.js
const _ = require('lodash');

const commands = require('./');

module.exports = {
    description: 'Displays a list of all available commands.',
    func: function(message) {
        const commandNames = _.sortBy(Object.keys(commands));
        const commandLengths = commandNames.map(n => n.length);
        const padSize = _.max(commandLengths) + 2;
        
        let res = 'Available commands:```';
        
        commandNames.forEach(n => {
            res += `${_.padEnd(n, padSize, ' ')}${commands[n].description}\n`;
        });
        
        res += '```';
        
        message.channel.send(res);
    }
};

File: WhiffBot/Commands/Commands.cs
using Discord.WebSocket;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using WhiffBot.Data;

namespace WhiffBot.Commands
{
    public class Commands : ICommand
    {
        private IGuildRepository GuildRepo { get; set; }
        private IEnumerable<ICommand> _commands;

        public string Name => "commands";
        public string Description => "Sends all available commands to the user.";

        public Commands(IGuildRepository guildRepo, IEnumerable<ICommand> commands)
        {
            _commands = commands;
            GuildRepo = guildRepo;
        }

        public async Task Run(SocketMessage message, string args)
        {
            var prefix = "";
            var guildChannel = message.Channel as SocketGuildChannel;
            if (guildChannel != null)
                prefix = GuildRepo.Get(guildChannel.Guild.Id).Settings.Prefix.ToString();

            var nameLength = _commands.Select(c => c.Name.Length).Max();

            var res = "```";

            foreach (var command in _commands)
            {
                res += $"{prefix}{command.Name.PadRight(nameLength)}  {command.Description}\n";
            }

            res += "```";

            await message.Channel.SendMessageAsync(res);
        }
    }
}
